### PR TITLE
[[ AndroidAudioRecorder ]] Add android audio recorder library

### PR DESCRIPTION
**Pull request by livecodeali**
This patch adds an LCB library which gives access to the Android
MediaRecorder API, allowing audio recording on Android. It also
includes a sample stack demonstrating the feature.

### CODE REVIEW

File: extensions/libraries/androidaudiorecorder/androidaudiorecorder.lcb [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    -- name, (i.e. /storage/emulated/0/pFileName.mp4 or equivalent)
+
+    CreateRecorder()
+    
+    if mRecorder is nothing then
+        throw "unable to create recorder"
+    end if
+    
+    variable tFile as JString
+    put GetAbsolutePathToPublicDocument(pFileName) into tFile
```
Comment (runrevmark) [inline extensions/libraries/androidaudiorecorder/androidaudiorecorder.lcb]: Is there a reason we restrict this to external storage? What if the incoming path is absolute already?

**Comment by runrevmark** [comment]
My only query is about the 'filename' which is passed - we are automatically prepending the external storage location. I wonder if we should just make the filename be processed as it would normally for engine commands/functions which take filenames?

File: extensions/libraries/androidaudiorecorder/androidaudiorecorder.lcb (hunk 1 above)
Comment (livecodeali) [reply extensions/libraries/androidaudiorecorder/androidaudiorecorder.lcb]: Heh, good point... actually external storage path should probably be in an android utility library

Comment (runrevmark) [reply extensions/libraries/androidaudiorecorder/androidaudiorecorder.lcb]: As long as the external storage path is accessible via specialFolderPath() and the recorder functions report a suitable error if the file can't be created; we can probably defer any such processing to user script.

Comment (livecodeali) [reply extensions/libraries/androidaudiorecorder/androidaudiorecorder.lcb]: I don't think the external storage path is accessible via specialFolderPath...

Comment (runrevmark) [reply extensions/libraries/androidaudiorecorder/androidaudiorecorder.lcb]: I think it should be:
```
	static final String s_external_prefix = "external ";
	public String getSpecialFolderPath(String p_name)
	{
		boolean t_external = false;
		if (p_name.startsWith(s_external_prefix))
		{
			t_external = true;
			p_name = p_name.substring(s_external_prefix.length());
		}

		try
		{
			if (p_name.equalsIgnoreCase("documents"))
			{
				if (t_external)
					return getContext().getExternalFilesDir(null).getAbsolutePath();
				else
					return getContext().getFilesDir().getAbsolutePath();
			}
			else if (p_name.equalsIgnoreCase("temporary") || p_name.equalsIgnoreCase("cache"))
			{
				if (t_external)
					return getContext().getExternalCacheDir().getAbsolutePath();
				else
					return getContext().getCacheDir().getAbsolutePath();
			}
		}
		catch (Exception e)
		{
			return "";
		}

		return "";
	}
```
Via 'external documents' - maybe? I think that's the same call you are using in LCB.

Comment (livecodeali) [reply extensions/libraries/androidaudiorecorder/androidaudiorecorder.lcb]: Ah! It is not documented!

**Comment by runrevmark** [comment]
@livecode-vulcan review ok be9d708

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok be9d708360d53ae0825ce256d37f2f9f97ac4ad6

**Comment by livecode-vulcan** [comment]
😎 test success be9d708360d53ae0825ce256d37f2f9f97ac4ad6

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/767)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/783)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/757)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/701)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/704)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/761)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/783)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/766)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/783)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/765)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/776)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/797)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/761)

**Pull request merged**